solc/CommandLineParser.cpp

```diff
@@ -582,7 +583,12 @@ General Information)").c_str(),
 	outputFormatting.add_options()
 		(
 			g_strPrettyJson.c_str(),
-			"Output JSON in pretty format. Currently it only works with the combined JSON output."
+			"Output JSON in pretty format."
+		)
+		(
+			g_strPrettyJsonIndent.c_str(),
```

Comment:
What name do you suggest? Something like '-ppi'? There is only one short option so far (-o).

Reply:
Short options must be a single letter. How about `-P`?

Reply:
Do we really need short options? How many do we have?

Reply:
As far as I know, only one (-o).

Reply:
We don't **need** them but typing `--pretty-print --pretty-print-indent 4` every time is not great. Which is why my initial proposal was to just have `--pretty-print 4` instead.

This is meant for CLI use so making it actually convenient would be nice.

Reply:
I suppose this features will most exclusively used on CLI (given the returned JSON can already be formatted in whatever way), so wouldn't a single short option with an optional parameter work the best?

Reply:
It would, though I don't see a reason not to provide a long option as well. It's useful when you're giving a command to someone or using it in a script and don't want it to be too cryptic.

Basically being able to write `solc -P` or `solc -P 4` or `solc --pretty-print` or `solc --pretty-print=4` would be best IMO and it's pretty straighforward to achieve with `boost::program_options`. Just a matter of agreeing on that.

I suggested it in https://github.com/ethereum/solidity/pull/11625#issuecomment-875042486 but @sladecek went with a separate option to make it more extensible in case more options need to be added later. I liked my proposal better but did not mind a separate option as long as a short name is available.

Reply:
OK, I will change that. ``-P`` will be a short form of `--pretty-print` and ``--pretty-print-indent`` will be removed and replaced by an optional parameter.

Reply:
By the way, I suggested `-P` rather than `-p` mostly because `--pretty-print` is a somewhat peripheral option and I thought we might want to save the name for a more important option in the future. From what @axic said, I guess we'll probably not be adding short versions for other options though so using `-p` would be fine too.

Reply:
I see a problem: there is an ambiguity with positional parameters. ``--pretty-print 5`` can mean either pretty printing with the indent of 5 or ``--pretty-print`` with default indent and input file named ``5``.

Reply:
It's fine if it's only interpreted as option value when `=` is used. You can get that behavior with `implicit_value()`. See [What is the difference between default_value and implicit_value in boost::program_options?](https://stackoverflow.com/questions/2829806/what-is-the-difference-between-default-value-and-implicit-value-in-boostprogra)

Reply:
`implicit_value` is OK, but I cannot find a way to enforce the equal sign. 

When the indent value '2' is present, it  works as described in the article:

> build/solc/solc --pretty-json 2  input.sol
> "input.sol" is not found.

or

> build/solc/solc --pretty-json=2  input.sol
> "input.sol" is not found.

But without the number '2' the following positional parameter cannot be distinguished (imagine input file named '2').

>build/solc/solc --pretty-json input.sol
> the argument ('input.sol') for option '--pretty-json' is invalid

double dash ``--`` helps

> build/solc/solc --pretty-json -- input.sol
> "input.sol" is not found.

Reply:
Well, looks like the situation is pretty ugly. In version 1.59 boost introduced a breaking change to make `implicit_value()` require `=` but people really did not like it and in 1.64 they restored the old behavior (https://github.com/boostorg/program_options/issues/25). We could get the old behavior by redefining `implicit_value()` but I'd rather avoid complicating things like that.

I think that the new behavior is not that bad actually, we could live with that, but the problem is that it might break scripts that just happen to have a file name after `--pretty-json` that used to be perfectly valid. They would only require moving the argument but it's still more breaking than I'd like. Fortunately we require boost >= 1.65 because that would be additional complication - we would have to bump the minimal version to ensure no one gets inconsistent behavior just by linking with some version between 1.59 and 1.64.

Looks like just giving `--pretty-print-indent` a short option name involves much fewer complications so I'd go with that.

Reply:
I added ``-P`` option and a test.

Reply:
I just got feedback from @axic and he's still strongly objecting to adding more short options and he has a point:

> The one character options are terrible in the sense you run out of them fairly quickly, hard to remember many of them, and knowing how long it takes removing deprecated ones it is a large baggage/legacy to carry.
> Let's say you choose -I for the short form, then when we want to introduce  something where -I makes more sense. Then we either 1) wait for a brraking release, 2) choose a worse option.
> Well it is not easy choosing a good short one incrementally :)

After discussing it we came to a conclusion that renaming `--pretty-print-indent` to `--json-indent` and making it enable `--pretty-print` automatically (instead of reporting an error) would be an acceptable alternative.

Sorry about bothering you about this again at this point, but could you change it so that it works like this?

Reply:
Sure, no problem.

Reply:
Finished.